# NAT-4
Identify Indels in Natera Whole Exome Data

## Methods

The indels pipeline was modified as follows.

It takes as input a pair of matched normal/tumor BAM files.

It extracts from the BAM files all mapping that overlap with an amplicon of the
manifest by at least 20 base pairs. From each such mapping it extracts the
corresponding read and complete it by the amplicon sequence in order the resulting
sequence covers the full amplicon. The assumption underlying this operation is
that the added sequence is not expected to induce an indel call as it will align
perfectly with the amplicon. Then for each such read it creates identical forward
and reverse reads and writes them into FASTQ files. This results in two annotated
FASTQ files similar in format to the ones used in the pipeline usually, with the
addition that the annotation of each read also includes a field specifying the
exact overlap between the original read and the amplicon; this annotation is not
used in the pipeline but could be useful to compute more precisely the VAF of
called variants. Another important point is that reads that overlap two amplicons
will then be duplicated and be associated to both amplicons.

From there, the indels pipeline is used as usual. This implies that some of the
generated read pairs will not result in merged reads if the reads have a mutation
in one of the amplicon primers.

This pipeline is implemented in the branch `NAT-4` of the indels pipeline github
repo `https://github.com/contextual-genomics/indels-pipeline`.

Variants with a VAF below 0.375%% were not called.

## Data

The data used are the BAM files generated by our alignment available at
`s3://dnanexus-ch?prefix=altera-pilot-study-canexia/alignments/`.

The directory `data` contains the amplicons manifest
`CG001v5.1_Amplicon_Manifest_Panel5.1.14_20201119.tsv` as well as the expected
variants. For a given tumor sample, all expected indels are in the file
`<SAMPLE>.vcf` while the indels whose position falls within an amplicon of our
manifest are in the file
`<SAMPLE>_CG001v5.1_Amplicon_Manifest_Panel5.1.14_20201119.vcf`.
The set of all expected indels is aggregated in the file `ALL_SAMPLES_expected.vcf`.

## Results

The results are available in the directory `results`. For each sample, this
directory contains an unannotated VCF file of all indels observed in the
alignments generated by the pipeline (files `*_indels_features_unfiltered.vcf`)
and the called indels annotated by snpEFF (files `*indels_filtered_snpeff.vcf`).

The set of all called indels is in the file `ALL_SAMPLES_PREDICTED.vcf`.

All the expected indels are found by the modified pipeline, but for two that are
in MSI amplicons. All other found vaerants are either in samples for which we do 
nt have the VCF file either have a very low VAF (<1%).
